Implementing Bloom Annotator

ian-cho opened the pull request:
## Why are these changes needed?

Add an additional column to indicate whether the document is present in the pre-existing Bloom filter model. This is specifically intended for the GneissWeb release.

ian-cho committed:
push code only

ian-cho committed:
Add files via upload

ian-cho committed:
Add files via upload

ian-cho committed:
Add files via upload

ian-cho committed:
Add files via upload

ian-cho committed:
Merge branch 'IBM:dev' into dev

ian-cho committed:
Update README.md

ian-cho committed:
Add files via upload

ian-cho committed:
Update local.py

ian-cho committed:
Update local_python.py

ian-cho committed:
Update test_bloom.py

ian-cho committed:
Update README.md

ian-cho committed:
Update README.md

shahrokhDaijavad commented:
@ian-cho I reviewed the README file and ran `python local_python.py` successfully on my laptop. Please add details to issue #981. 
Also, as done by all transforms, please add a simple notebook for the Python run-time (and a link to it from the README file).

ian-cho committed:
Add files via upload

ian-cho committed:
Merge branch 'IBM:dev' into dev

ian-cho committed:
Update README.md

ian-cho commented:
@shahrokhDaijavad @touma-I added notebook for python runtime and also put a link to it from README file  https://github.com/ian-cho/data-prep-kit/blob/dev/transforms/universal/bloom/bloom_python.ipynb

shahrokhDaijavad commented:
@ian-cho I haven't had time to test the notebook, but I just realized that you are using bf.bloom model. Is that open source and available to download, so we don't need to include it in the repo?

ian-cho commented:
@shahrokhDaijavad thanks for pointing this out. Currently, I am not aware of any small, downloadable models okay for this demonstration. Thus, I trained a small model locally. In the future, we can replace this local model path with a Hugging Face model when it is ready later I guess.

shahrokhDaijavad commented:
> @shahrokhDaijavad thanks for pointing this out. Currently, I am not aware of any small, downloadable models okay for this demonstration. Thus, I trained a small model locally. In the future, we can replace this local model path with a Hugging Face model when it is ready later I guess.

Thank you, @ian-cho. The size of the model 240KB is quite small, but for us to include it statically in an IBM-owned open repo, we need permission from IBM. I will bring this up with Bhatta, who may be able to get permission for us.

shahrokhDaijavad commented:
@BishwaBhatta  As you may know, @ian-cho is using a small model that he has trained himself for this transform.  This is fine for testing, but we cannot merge the code with that model included in the outer repo. So, if this small model will not be used at the end and an HF model will replace it, after successful testing, should we take it out before merging the code?

ian-cho commented:
@shahrokhDaijavad Hi, thanks for asking.

-  I tested the _gneissweb.bloom_ (28GB) that Yohei trained, the current bloom transform code work with it. So, from my side, it is okay to remove that small model and move forward. But I am open to @BishwaBhatta's opinion.
- @BishwaBhatta asked if the code can process fineweb parquet which has more than 30 columns. Among others, the document column is called _text_ not _contents_, I think we have been through this naming issue for HAP, so shall we stick to the name _contents_ instead of _text_ as in [test1.parquet](https://github.com/ian-cho/data-prep-kit/blob/dev/transforms/universal/bloom/test-data/input/test1.parquet)? Otherwise, I would like to upload new test1.parquet.
Thank you!

shahrokhDaijavad commented:
> @shahrokhDaijavad Hi, thanks for asking.
> 
>     * I tested the _gneissweb.bloom_ (28GB) that Yohei trained, the current bloom transform code work with it. So, from my side, it is okay to remove that small model and move forward. But I am open to @BishwaBhatta's opinion.
> 
>     * @BishwaBhatta asked if the code can process fineweb parquet which has more than 30 columns. Among others, the document column is called _text_ not _contents_, I think we have been through this naming issue for HAP, so shall we stick to the name _contents_ instead of _text_ as in [test1.parquet](https://github.com/ian-cho/data-prep-kit/blob/dev/transforms/universal/bloom/test-data/input/test1.parquet)? Otherwise, I would like to upload new test1.parquet.
>       Thank you!

Thank you, @ian-cho, for testing with the large model. For consistency with everything else in DPK, I think we should stick with the _contents_ instead of _text_ for the document column.

ian-cho commented:
@shahrokhDaijavad ok, then can I comment out this [line](https://github.com/ian-cho/data-prep-kit/blob/fe98d4095fc034b0aadbfea1422667f0c90c3d40/transforms/universal/bloom/dpk_bloom/transform.py#L67C7-L67C43) ? The transform does not need _text_ or _contents_. By doing this, we can stick to `contents` naming and also users do not have to change the name of fineweb parquet.

shahrokhDaijavad commented:
> @shahrokhDaijavad ok, then can I comment out this [line](https://github.com/ian-cho/data-prep-kit/blob/fe98d4095fc034b0aadbfea1422667f0c90c3d40/transforms/universal/bloom/dpk_bloom/transform.py#L67C7-L67C43) ? The transform does not need _text_ or _contents_. By doing this, we can stick to `contents` naming and also users do not have to change the name of fineweb parquet.

@touma-I  Do you see any problem with what @ian-cho is asking?

shahrokhDaijavad commented:
@ian-cho Sorry that I did not test the notebook until today. I did a `make venv` and in my `venv,` tried `python local_python` and it ran successfully. I had deleted the local output folder first (it should not have been put in the repo, as it gets created after the run, so please delete it). After the run, the output folder was created and had the 2 `metadata.json `and `test1.parquet `files were there. However, if I run the notebook, I get the error below:
<img width="898" alt="image" src="https://github.com/user-attachments/assets/57b367ed-8436-4ed9-99f1-3ffbe7f0269a" />
What am I missing? Can you please look into this? Thanks.

ian-cho committed:
Delete transforms/universal/bloom/output directory

ian-cho committed:
Merge branch 'IBM:dev' into dev

ian-cho commented:
Hi, @shahrokhDaijavad, thank you for trying!

> it should not have been put in the repo, as it gets created after the run, so please delete it

Sure, deleted.

> What am I missing? Can you please look into this? 

It was the path issue on my end. In your screenshot, please uncomment the 18th line and comment out the 19th line to see if works.
<img width="770" alt="Screenshot 2025-02-03 at 13 00 46" src="https://github.com/user-attachments/assets/ee5be30f-378b-4d15-a75a-d89c16e96785" />

ian-cho committed:
Update transform_python.py

shahrokhDaijavad committed:
Same chnage needed in 2 more files

Signed-off-by: SHAHROKH DAIJAVAD <shahrokh@us.ibm.com>

shahrokhDaijavad commented:
Thanks, @ian-cho for pointing out the problem. I pushed the same change in 2 other files, and now the notebook works well. Thanks for deleting the output folder.
Any idea why the workflow test fails in CI/CD? If not, I will discuss it with @touma-I tomorrow.

ian-cho commented:
Hi, @shahrokhDaijavad, I have no ideas why it failed. Your and @touma-I 's help would be appreciated!

touma-I committed:
added CI/CD workflow

Signed-off-by: Maroun Touma <touma@us.ibm.com>

shahrokhDaijavad committed:
Fix for the test-src failing in CI/CD

Signed-off-by: SHAHROKH DAIJAVAD <shahrokh@us.ibm.com>

shahrokhDaijavad commented:
@ian-cho After @touma-I fixed the workflow test failure, I saw that test-src was failing and I pushed the same fix that we did for using the new API and now everything is passing in CI/CD.
I think the only remaining issue is that we still have the small `bf.bloom` file included. If this is a meaningless model that is just for testing, it may be ok keeping it (like small meaningless parquet files we use for testing).

ian-cho committed:
Update transform.py

ian-cho commented:
@touma-I @shahrokhDaijavad Thanks for your fix! Retaining bf.bloom would make the repo self-contained if it is fine to keep it. Also, I added two comments [here](https://github.com/ian-cho/data-prep-kit/blob/dev/transforms/universal/bloom/dpk_bloom/transform.py#L67) and [here](https://github.com/ian-cho/data-prep-kit/blob/dev/transforms/universal/bloom/dpk_bloom/transform.py#L71) for the user's convenience for filtering out FineWeb into GneissWeb if it is okay.

shahrokhDaijavad commented:
Thank you, @ian-cho. Some gneissweb fasttext models were added to HuggngFace yesterday. Who is adding the `gneissweb.bloom `(28GB) that Yohei trained to HF? 
I am approving this PR and pass it to @touma-I, before it is merged.

shahrokhDaijavad reviewed (approved):
LGTM.

shahrokhDaijavad committed:
Improved README file

Signed-off-by: SHAHROKH DAIJAVAD <shahrokh@us.ibm.com>

shahrokhDaijavad commented:
@ian-cho As I was going over the README file with @touma-I today, we felt that it was not clear enough for our audience in the open-source community why we are doing this. So, I made some changes in the README file. Can you please review my changes and see if you want to modify/add anything? Thanks.

ian-cho commented:
Hi, @shahrokhDaijavad @touma-I , thanks. The added summary looks good to me.

ian-cho committed:
Merge branch 'IBM:dev' into dev

ian-cho committed:
Update transform.py

ian-cho committed:
Update local_python.py

ian-cho committed:
Update test_bloom.py

ian-cho committed:
Delete transforms/universal/bloom/bf.bloom

ian-cho committed:
Update transform.py

ian-cho committed:
Update transform.py

ian-cho committed:
Update requirements.txt

ian-cho commented:
@touma-I @shahrokhDaijavad @BishwaBhatta (@issei-ibm) I updated the model path in the scripts to ibm-granite/GneissWeb.bloom. However, during on-the-fly testing, I run into the OOM error due to the model's 28GB size. any ideas? Thanks a lot!

shahrokhDaijavad commented:
Thank you for trying this out, @ian-cho. I am not surprised, because I would definitely get an OOM on my laptop that has 16GB memory. How much memory do you have on your laptop? When you said a couple of weeks ago (see above) that you tried it with the 28GB model that Yohei trained, how did you do it? 
In any case, if we have some idea of the minimum memory needed to try this successfully when the 28GB file is downloaded on the fly, we should put that in the README file. Most community users though won't have access to machines with more than 16GB memory. Maybe you should put back the small bf.bloom so that local testing succeeds, but in the README, we point and link to the real model file on HF/IBM-granite.

ian-cho commented:
@shahrokhDaijavad No, on my laptop, it was successful to load 28GB model with the on-the-fly test. When I push it to DPK, one test failed due to OOM error as follow: 
![Screenshot 2025-02-15 at 10 02 12](https://github.com/user-attachments/assets/92d595d9-f15e-41c7-b499-38344b5a6373)
See from 497th line.

shahrokhDaijavad commented:
Great. Thank you @ian-cho. So, the OOM happens in the Ubuntu server that runs CI/CD tests. Please keep the code as is with the real model.  If your laptop has more than 16GB of memory, please let me know, and I will add a sentence to the README file, stating that "using the large model file will require a machine with more than 16GB of memory".

ian-cho commented:
@shahrokhDaijavad Hi, Thanks. My laptop has more than 16GB memory.

shahrokhDaijavad committed:
README and notebook changes

Signed-off-by: SHAHROKH DAIJAVAD <shahrokh@us.ibm.com>

shahrokhDaijavad commented:
@ian-cho Can you please review the changes I made to the README and to the notebook? Thanks.

code4nlp1713 commented:
Hi, @shahrokhDaijavad, thank you for making these changes. It looks good to me. Just two quick questions:
- Is laptop with 16GB sufficient to load 28GB bloom filter model? (Mine actually has 64GB memory)
- As for [bloom_python.ipynb](https://github.com/IBM/data-prep-kit/pull/978/commits/20b348461a1f2359da55f6a1b4c685ec299d356a#diff-e91971a8dd5613cfed9467fce2d461ce4745f4d289effe350efd7fb87b19765a), only the OOM error remains, while everything else seems fine after the modification? Thanks a lot.

ian-cho commented:
Hi, @shahrokhDaijavad, thank you for making these changes. It looks good to me. Just two quick questions:
- Is laptop with 16GB sufficient to load 28GB bloom filter model? (Mine actually has 64GB memory)
- As for [bloom_python.ipynb](https://github.com/IBM/data-prep-kit/pull/978/commits/20b348461a1f2359da55f6a1b4c685ec299d356a#diff-e91971a8dd5613cfed9467fce2d461ce4745f4d289effe350efd7fb87b19765a), only the OOM error remains, while everything else seems fine after the modification? Thanks a lot.

shahrokhDaijavad committed:
comment cell for notebook shows the new way for pip install

Signed-off-by: SHAHROKH DAIJAVAD <shahrokh@us.ibm.com>

shahrokhDaijavad commented:
@ian-cho I made a small change in the comment cell of the notebook that mentions pip install. With the new APIs, we don't need to pip install data-prep-toolkit anymore. 
As for your questions, in my laptop that has only 16GB of memory, when I tried today, I got a Jupyter kernel crash when I ran the notebook, so 16GB is not enough. 
For your second question, yes, I am fine with approving this PR (even with the CI/CD error), but I am not able to merge and Maroun has asked me not to ask other admins who can merge until he comes back next week. Is there an urgent need to merge this before next week? 
I am able to create a pip install release on PyPi that includes the bloom transform, even if it is still in PR and not merged. Is there an urgent need for that?

ian-cho commented:
@shahrokhDaijavad Thank you for the modification. In HuggingFace’s release of the GinessWeb Bloom filter, the link to DPK now points to the dev branch of my personal Git repository. It should be pointing to the DPK outer branch after the merge.

Understood, @ian-cho We will change the link after the merge.

ian-cho committed:
Merge branch 'IBM:dev' into dev

ian-cho committed:
Add files via upload

Added a huggingface-based bloom filter usage notebook. Here is the reason why we did it: this enhancement will allow for data to be read on the fly from Hugging Face API without the user having to download the data to a local machine/server. Only the final annotated version will have to be saved to disk.  Also the Hugging Face API uses parquet file format.  Where as when one downloads Fineweb it could be of other formats like json etc.  So reading from Hugging Face API would make it very compatible.

ian-cho commented:
@touma-I @shahrokhDaijavad Hi, for this PR, we added a new HuggingFace-based usage notebook, `bloom_python_HuggingFace.ipynb`, here is our justification (@BishwaBhatta): This enhancement will allow for data to be read on the fly from Hugging Face API without the user having to download the data to a local machine/server. Only the final annotated version will have to be saved to disk.  Also the Hugging Face API uses parquet file format.  Where as when one downloads Fineweb it could be of other formats like json etc.  So reading from Hugging Face API would make it very compatible.

Please let me know if this enhancement is fine with you and if so, shall we keep both notebooks or replace? Thank you!

cc. @issei-ibm

shahrokhDaijavad commented:
> @touma-I @shahrokhDaijavad Hi, for this PR, we added a new HuggingFace-based usage notebook, `bloom_python_HuggingFace.ipynb`, here is our justification (@BishwaBhatta): This enhancement will allow for data to be read on the fly from Hugging Face API without the user having to download the data to a local machine/server. Only the final annotated version will have to be saved to disk. Also the Hugging Face API uses parquet file format. Where as when one downloads Fineweb it could be of other formats like json etc. So reading from Hugging Face API would make it very compatible.
> 
> Please let me know if this enhancement is fine with you and if so, shall we keep both notebooks or replace? Thank you!
> 
> cc. @issei-ibm

Thank you, @ian-cho. I like this enhancement! My thinking is to delete the other notebook and just keep this one. The README file also needs to be changed. All three comment sections you have added to this notebook (especially the one about configuration parameters) should also be added to the README file. @touma-I, Do you see any benefit in keeping the old notebook?

shahrokhDaijavad commented:
@ian-cho I have discussed this transform today with @touma-I, and our consensus is: 
1) Your enhanced notebook is the ultimate guide for the new GneissWeb-interested users of DPK! 
2) Please keep the old notebook, but the old notebook and all the other code need to go back and use the small model (240K file size) that you had created before, so the small "toy" model will be in the repo. This is like the toy parquet test files we use everywhere. This way, we won't run into OOM problems on machines with 16GB memory in unit tests locally or on the CI/CD server. Everything other than the new notebook uses that local model. 
3) On the issue of credentials for HF, @touma-I has implemented a solution in a new PR for lang_id transform (PR #1068) (uses environment variable HF_READ_ACCESS_TOKEN) and this will be the solution we need to adopt for all transforms that download models from HF. 

On point 3, @touma-I just wrote to @ran-iwamoto on her transform:

Please see how the HF credential was fixed for lang_id https://github.com/IBM/data-prep-kit/pull/1068 and follow the same steps to fix it here. Mainly, in the git repo we define a secret HF_READ_ACCESS_TOKEN that contains the credential, then in the workflow .github/workflows/test-language-lang_id.yml lines 72 and 73 we expose the secret as part of the environment variable and then in the python code we use os.environ.get() to get the secret from the environment variable . We can have a quick call if things are not clear after you have reviewed PR 1068. cc: @ran-iwamoto @issei-ibm

ian-cho committed:
Add files via upload

ian-cho committed:
Update bloom_python.ipynb

ian-cho committed:
Update bloom_python.ipynb

ian-cho committed:
Update local_python.py

ian-cho committed:
Update transform.py

ian-cho committed:
Update test_bloom.py

ian-cho committed:
Update test-universal-bloom.yml

ian-cho committed:
Update local.py

ian-cho committed:
Update test_bloom.py

ian-cho committed:
Update test_bloom.py

ian-cho committed:
Update local.py

ian-cho commented:
@shahrokhDaijavad @touma-I Thanks!

> Please keep the old notebook, but the old notebook and all the other code need to go back and use the small model (240K file size) that you had created before, so the small "toy" model will be in the repo.

Yes, I put the toy model back and modified all the code related to toy model, now I found OOM error has been resolved.

> On the issue of credentials for HF, @touma-I has implemented a solution in a new PR for lang_id transform (PR https://github.com/IBM/data-prep-kit/pull/1068) (uses environment variable HF_READ_ACCESS_TOKEN) and this will be the solution we need to adopt for all transforms that download models from HF.

I also added env path as @touma-I did for lang_id and added this to the code scripts. Now all checks passed. Please check if it is okay. Thanks! (cc. @issei-ibm)

shahrokhDaijavad commented:
Thank you very much, @ian-cho! Looks like you did everything, but I will look more closely tomorrow. I appreciate your fast action. 
Since this is fresh in your mind, can I please ask you for a favor? @ran-iwamoto just did the same solution for env path that @touma-I suggested in her PR here: #1046, and it is still failing the test-src due to HF credential. So, she must be missing one small step, and you may be able to help her. Thank you.

ian-cho committed:
Add files via upload

ian-cho commented:
@shahrokhDaijavad I checked @ran-iwamoto 's error case, mine seems to be slightly different from that where Ran's error seems to occur in the test .py file, whereas in my case, I do not need the HF model or its credentials in any code including test (I used a toy model) except for the new notebook, `bloom_python_HuggingFace.ipynb`. Your closer look and check would be appreciated!

touma-I commented on transforms/universal/bloom/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+rbloom==1.5.2
+pandas==2.2.2
```
Can we relax any or all of those requirements? Ideally:
pandas should not have a tag but use the latest
How about others? Do we need a tag and if so can you use >=  ?

ian-cho replied:
fix this by replacing `==` with `>=`

ian-cho committed:
Merge branch 'IBM:dev' into dev

ian-cho committed:
Update requirements.txt

relax version number in the requirements.txt for bloom filter annotator

ian-cho committed:
Update pyproject.toml

added transform for bloom filter annotator

ian-cho committed:
Update pyproject.toml

fix a typo in the previous update

ian-cho committed:
Update pyproject.toml

fix a type in the previous update

touma-I commented on transforms/pyproject.toml:
```diff
@@ -83,7 +84,7 @@ language = { file = [
 "universal/ededup/requirements.txt",
 "universal/fdedup/requirements.txt",
-"universal/hap/requirements.txt",
+"universal//requirements.txt",
```
Something is missing here!

ian-cho replied:
fixed it.

touma-I commented on transforms/pyproject.toml:
```diff
@@ -83,7 +84,7 @@ language = { file = [
 "universal/doc_id/requirements.txt",
 "universal/ededup/requirements.txt",
 "universal/fdedup/requirements.txt",
-"universal/hap/requirements.txt",
```
Maybe deleted by accident?!

ian-cho replied:
oh, fixed!

touma-I commented on transforms/universal/bloom/bloom_python_HuggingFace.ipynb:
This notebook is failing with the following error below. since this transform is still not in pypi, I commented out the first cell with  #%pip install 'data-prep-toolkit-transforms[bloom]' and setup the environment using `make venv` as described below:
 
```
cd transforms/universal/bloom
make venv
source venv/bin/activate
pip install Jupyter lab
Jupyter lab
```
Then ran the notebook until I get to this error:

<img width="1007" alt="Screenshot 2025-03-06 at 12 10 39 PM" src="https://github.com/user-attachments/assets/fe322ede-0df3-492e-9e2a-ea7955616cc3" />

shahrokhDaijavad replied:
@touma-I @ian-cho I tried and got the same error as above.

ian-cho replied:
@touma-I @shahrokhDaijavad Thank you for testing! Your feedback is very valuable. The issue may be caused by the BLOOM_MODEL used when instantiating the BLOOMTransform class. I have modified `bloom_python_HuggingFace.ipynb` by adding model_path. Please test it, many thanks!

touma-I reviewed (commented):
Seem to be getting an error with the notebook.

ian-cho committed:
Update pyproject.toml

fixed the unexpected deletion

ian-cho committed:
Merge branch 'IBM:dev' into dev

ian-cho committed:
Add files via upload

shahrokhDaijavad commented:
Thank you for your fix @ian-cho. I now pass the point of error above, but the Jupyter kernel crashes because of low memory/storage. 
@touma-I I think this will now work on your machine.
